### ISSUE TITLE
Double (denested) sum does not doit()

### ISSUE THREAD

**Issue by oscargus** [sympy/sympy#15767]
(May be related to #5790)

There is an inconsistency (or maybe I do not get something) when evaluating the same sum written in two different ways:

    [k, l] = symbols('k l', cls=Idx)
    g = IndexedBase('g')
    K = symbols('K', type=Integer, real=True, positive=True)

    sum1 = summation(g[k]*summation(g[l],(l,k+1,K)),(k,0,K))
    sum2 = summation(summation(g[k]*g[l],(l,k+1,K)),(k,0,K))
    sum3 = summation(g[k]*g[l],(l,k+1,K),(k,0,K))
   
Now `sum1.subs(K,2).doit()` gives the expected

    (g[1] + g[2])⋅g[0] + g[1]⋅g[2]

while `sum2.subs(K,2).doit()` and `sum3.subs(K,2).doit()` gives

      2       2              
     ___     ___             
     ╲       ╲               
      ╲       ╲     g[k]⋅g[l]
      ╱       ╱              
     ╱       ╱               
     ‾‾‾     ‾‾‾             
    k = 0 l = k + 1          

I have also noted that `k`gets the correct value in the second sum if we add something using it and that as long as `k` is used before the second summation in `sum2` it expands as expected. For example, multiplying with `k`. 

(In a sudden flash of inspiration I also tried out 

    sum4 = summation(g[k]*g[l],(k,0,K),(l,k+1,K))

but that gives the incorrect

        2                                        
       ___                                       
       ╲                                         
        ╲     (g[0]⋅g[l] + g[1]⋅g[l] + g[2]⋅g[l])
        ╱                                        
       ╱                                         
       ‾‾‾                                       
    l = k + 1  

so that can be left out as as error source.)

**Comment by oscargus** [comment]
An observation while trying to get a bit more insight (when evaluating `sum2`):

The result from sympy/concrete/summations.py(187) `dif = b - a` is an integer, while the result from the same operation with what should be the same operands from sympy/concrete/summations.py(941) is not. (The result is `-k + 1` in both cases.)

Not sure that this causes the problem though. It rather seems like the fact that `k` does not have a value when evaluating the "inner" one of the denested summations causes the issue. Hence, I guess one way would be to check if the bounds of a summation depends on other summation indices and in that case do something different (reorder the evaluation in some clever way?).

**Comment by oscarbenjamin** [comment]
Here's a simpler way to reproduce:
```python
i, j = symbols('i j', cls=Idx)
g = IndexedBase('g')

sum1 = summation(g[i], (i,0,j), (j,0,2))
pprint(sum1.doit())
```
Output:
```
  2     j       
 ___   ___      
 ╲     ╲        
  ╲     ╲   g[i]
  ╱     ╱       
 ╱     ╱        
 ‾‾‾   ‾‾‾      
j = 0 i = 0  
```
One way to deal with this might be to expand the outer summation into an sum of summations and then call doit on each sub-summation. (Maybe that is what happens for `sum1`)

**Comment by oscargus** [comment]
Yes, probably the way to approach it is to analyze the possible relation between summation bounds and based on that make a decision how to proceed. (My actual case involves four summations which are "pairwise" related, so like the top but with a similar sum inside. Which should be possible to solve in two stages: first solve the inner sum and then throw the result back on the summation function again.) As far as I can tell there is no checking of bounds depending on other summation variables at all.

For the cases that works, the additional variable before the second sum makes it stay as two separate sums, so it is somehow related to denested sums/multiple bounds depent on other indices in the same `Sum`.

**Comment by oscarbenjamin** [comment]
The current implementation works from the innermost sum out. For any given sum if the limits aren't integers and it's not possible to find an analytic formula then it does nothing:
[https://github.com/sympy/sympy/blob/sympy-1.3/sympy/concrete/summations.py#L942](https://github.com/sympy/sympy/blob/sympy-1.3/sympy/concrete/summations.py#L942)

I think it should *always* be possible to implement `doit` working in from the outermost sum. It would be less efficient that way for some cases though since you'd repeatedly .doit on the inner summations that may not change.

**Comment by oscargus** [comment]
I agree that it should always be possible to evaluate, at least explicitly, from the outermost sum.

This seems to (primarily) be an issue when dealing with `IndexedBase` and `Function` as the following experiments illustrate:

<pre>
In [4]: from sympy import *

In [5]: i,j,k = symbols('i j k', cls=Idx)

In [6]: summation(i, (k, 0, j), (j, i, 2), (i, 0, 2))
Out[6]: 
11

In [7]: summation(j, (k, 0, j), (j, i, 2), (i, 0, 2))
Out[7]: 
22

In [8]: summation(k, (k, 0, j), (j, i, 2), (i, 0, 2))
Out[8]: 
11

In [9]: summation(1, (k, 0, j), (j, i, 2), (i, 0, 2))
Out[9]: 
14

In [10]: g = IndexedBase('g')

In [11]: summation(g[i], (k, 0, j), (j, i, 2), (i, 0, 2))
Out[11]: 
6*g[0] + 5*g[1] + 3*g[2]

In [12]: summation(g[j], (k, 0, j), (j, i, 2), (i, 0, 2))
Out[12]: 
Sum((j + 1)*g[j], (j, i, 2), (i, 0, 2))

In [13]: summation(g[k], (k, 0, j), (j, i, 2), (i, 0, 2))
Out[13]: 
Sum(g[k], (k, 0, j), (j, i, 2), (i, 0, 2))
</pre>

The same results are obtained if `g` is a function.

Hence, it seems like if a bound on an "innermore" index is defined by an "outermore" index and that index is used for a `Function` or `IndexedBase`, the evaluation of these indices should be made in reverse order. It probably also makes sense to "disconnect" summations that does not depend on each other using some heuristic ordering, think:

    summation(g[k]*g[c], (k, 0, i), (a, 0, b), (b, 0, c), (c, 0, 2), (i, 0, 2))

In that case it would make sense to evaluate the indices `[k, i]` and `[a, b, c]` separately and check for possible reordering for each set of indices.

For `IndexedBase` it is enough to check if the index is of class `Idx`. However, for `Function` it is not enough. Hence, one should probably check if the index is used as an argument to either `IndexedBase` or `Function` (or a subclass).

**Comment by oscargus** [comment]
The weird thing is that on line 187 the same expression as you linked to is an integer (not sure how to link to code directly), as stated in the first post. Still, I would assume that `dif < 100` would not evaluate to `True` in that case.
